### PR TITLE
Auto adding `#` to color from colordialog

### LINKED ISSUE #607: Manually adding hex color in color picker dialog doesn't pre-fill with hashtag when changing table cell border or background
## Are you reporting a feature or a bug?

Feature request

## [Check if the issue is already reported](http://docs.ckeditor.com/#!/guide/dev_issues_readme-section-avoid-duplicates)

Extracted from #565. See also #590.

## Provide detailed reproduction steps (if any)

1. Go to http://sdk.ckeditor.com/samples/fullpreset.html.
1. Right click into table cell.
1. Select `Cell -> Cell Properties`.
1. Press `Choose` button next to _Background Color_.
1. Type color value in input field (just above `Clear` button) in hexadecimal 6-digit format **without** leading `#`. E.g. (`ABC123`).
1. Click OK.
1. Click OK.

### Expected result

`#ABC123` color was applied as an cell background.

### Actual result

No color applied due to invalid hex (`ABC123`) value.

## Other details

* Browser: All
* OS: All
* CKEditor version: 4.7.1
* Installed CKEditor plugins: Full preset

### PR DESCRIPTION
**Pull request by msamsel**
## What is the purpose of this pull request?

New feature
## Does your PR contain necessary tests?

yes

### This PR contains

- [x] Unit tests
- [x] Manual tests

## What changes did you make?
There will be automatically added `#` to colour in `colordialog` plugin, if colour will be 3 or 6 hexadecimal digit.

---

Closes #607.

### CODE REVIEW

File: plugins/colordialog/dialogs/colordialog.js [hunk 1]
```diff
@@ -22,6 +22,14 @@ CKEDITOR.dialog.add( 'colordialog', function( editor ) {
 		html: '&nbsp;'
```
Comment (Comandeer) [inline plugins/colordialog/dialogs/colordialog.js]: What is the purpose of the changes in this file?

File: plugins/colorbutton/plugin.js [hunk 2]
```diff
@@ -203,6 +203,11 @@ CKEDITOR.plugins.add( 'colorbutton', {
 				total = colors.length + ( moreColorsEnabled ? 2 : 1 );
 
 			var clickFn = CKEDITOR.tools.addFunction( function applyColorStyle( color, type ) {
+				// #565
```
Comment (Comandeer) [inline plugins/colorbutton/plugin.js]: Please add short comment explaining the change.

File: plugins/colorbutton/plugin.js [hunk 3]
```diff
@@ -203,6 +203,11 @@ CKEDITOR.plugins.add( 'colorbutton', {
...
+				if ( /^[0-9a-f]{3}([0-9a-f]{3})?$/i.test( color ) ) {
+					color = '#' + color;
```
Comment (Comandeer) [inline plugins/colorbutton/plugin.js]: IMO the fix is placed in incorrect plugin. I'd move it into `colordialog` one, inside `onClose` handler of the dialog.

File: tests/plugins/colordialog/collorbuttondialogs.js [hunk 4]
```diff
@@ -0,0 +1,240 @@
+/* bender-tags: editor,unit */
```
Comment (Comandeer) [inline tests/plugins/colordialog/collorbuttondialogs.js]: There's no need for `unit` tag.

File: tests/plugins/colordialog/manual/autohashadding.md [hunk 5]
```diff
@@ -0,0 +1,14 @@
+@bender-tags: manual, tc, 4.8.0, feature, 565
```
Comment (Comandeer) [inline tests/plugins/colordialog/manual/autohashadding.md]: There's no need for `manual` tag.

File: plugins/colordialog/dialogs/colordialog.js (hunk 1 above)
Comment (msamsel) [inline plugins/colordialog/dialogs/colordialog.js]: Change allows on writing unit test for this bug-fix.

Generally writing test for it was very tricky. Problem appeared when `ok` button is pressed then method `clearHighlight()` is called. What throw an error because unit test doesn't highlight colour in dialog as user does.
So I just add if-statement which check if focus is defined and also rearrange variable declaration to pass JShint.

**Comment by mlewand** [comment]
Ok to sum it up:

The fix should land in the `colordialog` plugin rather than in color button.

There's no question to that, as first of all the enhancement should propagate to all fetarues using `colordialog`, see "Cell Properties" dialog:

Cell Properties

Generally speaking it's not `colorbutton`'s job to mess with color format (there are exception but I'll leave it for sake of simplicity).

This ticket will depend on #590, as AFAICS otherwise we'd have to duplicate this logic in `colorbutton` plugin.

File: tests/plugins/colordialog/colordialog.js [hunk 6]
```diff
@@ -0,0 +1,121 @@
+/* bender-tags: editor */
```
Comment (f1ames) [inline tests/plugins/colordialog/colordialog.js]: Remember to always put `unit` tag if it is an unit test.

File: tests/plugins/colordialog/manual/addhash.md [hunk 7]
```diff
@@ -0,0 +1,13 @@
+@bender-tags: tc, feature, colordialog, 565, 4.8.0
+@bender-ui: collapsed
+@bender-ckeditor-plugins: table,tabletools,colordialog,toolbar,wysiwygarea,sourcearea,contextmenu
```
Comment (f1ames) [inline tests/plugins/colordialog/manual/addhash.md]: Is `sourcearea` plugin really needed here?

File: tests/plugins/colordialog/manual/addhash.md [hunk 8]
```diff
@@ -0,0 +1,13 @@
...
+
+1. Right click into table cell.
+1. Select `Cell -> Cell Properties`.
+1. Choose background color.
+1. Type color in hexadecimal 6-digit format **without** leading `#`. E.g. (`ABC123`).
+1. Confirm it.
+1. Inseted color should contain additional `#` at the beginning.
```
Comment (f1ames) [inline tests/plugins/colordialog/manual/addhash.md]: Typo: Inseted -> Inse**r**ted

File: tests/plugins/colordialog/manual/addhash.md [hunk 9]
```diff
@@ -0,0 +1,13 @@
...
+1. Repeat above steps for color with hexadecimal 3-digits format. E.g. (`FFF`).
+
+Note: You can also check some wrong situations.
```
Comment (f1ames) [inline tests/plugins/colordialog/manual/addhash.md]: It may be not clear what is **wrong situation**. You could rephrase it like:
> You can also check with colors in different formats (rgb, text) or invalid hex strings. For such colors # should not be added.

File: tests/plugins/colordialog/manual/addhash.md [hunk 10]
```diff
@@ -0,0 +1,13 @@
+@bender-tags: tc, feature, colordialog, 565, 4.8.0
```
Comment (f1ames) [inline tests/plugins/colordialog/manual/addhash.md]: I think we do not use `feature` tag (unless it was added recently)?

File: tests/plugins/colordialog/colordialog.js [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+/* bender-ckeditor-plugins: colordialog,wysiwygarea */
+
+( function() {
+	'use strict';
+
+	bender.editor = {};
+
+	var test = {
```
Comment (f1ames) [inline tests/plugins/colordialog/colordialog.js]: There is no need to assign it to a variable, you can just write it as`bender.test( { ... } )`.

File: tests/plugins/colordialog/colordialog.js [hunk 12]
```diff
@@ -0,0 +1,121 @@
...
+			editor.getColorFromDialog( function( color ) {
+				resume( function() {
+					assert.areSame( 'rgb(10, 20, 30)', color );
+				} );
+			} );
+			wait();
+		},
+
+		'test colordialog does not add hash empty value': function() {
+			var editor = this.editor;
```
Comment (f1ames) [inline tests/plugins/colordialog/colordialog.js]: All tests have very similar body, maybe it could be extracted to a separate function like `assertColor( editor, initialColor, expectedColor )` and then each test will be a one liner, e.g. `assertColor( this.editor,  '123456', '#123456')`.

If you put the function inside `test` object you can even omit `editor` param because of the same scope so `this.editor` will be available there.

File: tests/plugins/colordialog/colordialog.js [hunk 13]
```diff
@@ -0,0 +1,121 @@
...
+
+			editor.getColorFromDialog( function( color ) {
+				resume( function() {
+					assert.areSame( '#FDE', color );
+				} );
+			} );
+			wait();
+		},
+
+		'test colordialog does not add hash 1 digit': function() {
```
Comment (f1ames) [inline tests/plugins/colordialog/colordialog.js]: This test fails on IE8.

File: tests/plugins/colordialog/colordialog.js (hunk 6 above)
Comment (f1ames) [inline tests/plugins/colordialog/colordialog.js]: So we discussed this once again, if unit tag is not needed (not used by bender) it should be skipped.

File: tests/plugins/colordialog/manual/addhash.md (hunk 10 above)
Comment (msamsel) [inline tests/plugins/colordialog/manual/addhash.md]: As it was discussed some time ago there should either `bug` or `feature` tag? Or does it change since then?

Comment (f1ames) [inline tests/plugins/colordialog/manual/addhash.md]: Maybe I'm not up to date - if it was discussed with @mlewand or @Comandeer in the last 2-3 weeks it means it should be that way.

File: tests/plugins/colordialog/manual/addhash.md (hunk 7 above)
Comment (msamsel) [inline tests/plugins/colordialog/manual/addhash.md]: It might be helpful when you wish to check what was included to attribute of tag. E.g. if there will be some invalid string or colour hard to distinguish from others, then checking source might be helpful.

File: tests/plugins/colordialog/manual/addhash.md (hunk 10 above)
Comment (msamsel) [inline tests/plugins/colordialog/manual/addhash.md]: It was discussed around 12 - 13 June on `ckeditor4-chat`. I made some kind of sum up for us as pinned post to ckeditor4-chat.

File: tests/plugins/colordialog/colordialog.js [hunk 14]
```diff
@@ -0,0 +1,56 @@
...
+
+			editor.getColorFromDialog( function( color ) {
+				resume( function() {
+					assert.areSame( outputColor, color );
+				} );
+			} );
+			wait();
+		},
+
+		'test colordialog add hash to colors 6 digits': function() {
```
Comment (f1ames) [inline tests/plugins/colordialog/colordialog.js]: I think test names would be more readable like:
* `test colordialog add hash to colors - 6 digits`
* `test colordialog add hash to colors: 6 digits`
* `test colordialog add hash to colors with 6 digits color`

File: tests/plugins/colordialog/colordialog.js [hunk 15]
```diff
@@ -0,0 +1,56 @@
...
+			this.assertColor( '123456', '#123456' );
+		},
+
+		'test colordialog add hash to colors 3 digits': function() {
+			this.assertColor( 'FDE', '#FDE' );
+		},
+
+		'test colordialog does not add hash 1 digit': function() {
+			// IE8 don't allow on totally wrong values of attributes
```
Comment (f1ames) [inline tests/plugins/colordialog/colordialog.js]: _IE8 doesn't allowing invalid values_ means the attribute will not be set? I think it might be more clear with something like:
`IE8 doesn't allow to use invalid color values.` or `IE8 doesn't set invalid color values.` ?

File: tests/plugins/colordialog/manual/addhash.md [hunk 16]
```diff
@@ -0,0 +1,13 @@
...
+@bender-ui: collapsed
+@bender-ckeditor-plugins: table,tabletools,colordialog,toolbar,wysiwygarea,sourcearea,contextmenu
+
+1. Right click into table cell.
+1. Select `Cell -> Cell Properties`.
+1. Choose background color.
```
Comment (f1ames) [inline tests/plugins/colordialog/manual/addhash.md]: Does it mean to click `Choose` button and then type color in a `colordialog`? It might be a little unclear, and typing color without hash in `Cell Properties` dialog does not add hash (and color when closing dialog) so it may seem it is broken in this case.
To make it more clear you may use: `Click Choose next to Background Color` or something similar.

File: plugins/colordialog/dialogs/colordialog.js (hunk 1 above)
Comment (msamsel) [inline plugins/colordialog/dialogs/colordialog.js]: Color dialog hasn't got any unit test earlier.
This change:
```JavaScript
if ( focused ) {
	focused.removeClass( focusedColorLightCls );
	focused.removeClass( focusedColorDarkCls );
}
```
Allows on making nice unit test which set up values but doesn't touch focused colour. So test don't throw up errors when only value exists without focusing it.
